python/ray/data/datasource/file_based_datasource.py

```diff
@@ -192,6 +191,7 @@ def _resolve_paths_and_filesystem(
         raise ValueError("Must provide at least one path.")
 
     if filesystem and not isinstance(filesystem, FileSystem):
+        import fsspec
```

Comment:
You might want to catch an `ImportError` here and log and/or raise a more informative error. From my comment on the audit doc:
>This actually shouldn't be needed unless you're using an fsspec filesystem. fsspec import could be moved here (https://github.com/ray-project/ray/blob/5be6bda4cf4bce9f21143289fbf3641578ba0b8d/python/ray/data/datasource/file_based_datasource.py#L194-L200), with import errors caught to let the user that they need to pass in either a pyarrow or fsspec filesystem, and if they're trying to pass in the latter, then they need to install fsspec (although it probably wouldn't be possible to pass in an fsspec filesystem without installing fsspec first, so we could probably just error at that point saying that they need to give a pyarrow or fsspec filesystem).

Something like this maybe?
```python
    if filesystem and not isinstance(filesystem, FileSystem):
        err_msg = ("The filesystem passed must either conform to pyarrow.fs.FileSystem, or "
                            f"fsspec.spec.AbstractFileSystem. The provided filesystem was: {filesystem}")
        try:
            import fsspec
        except ModuleNotFound:
            # fsspec not not found, assuming that the passed filesystem wasn't an fsspec filesystem.
            raise TypeError(err_msg) from None
        if not isinstance(filesystem, fsspec.spec.AbstractFileSystem):
            raise TypeError(err_msg)